docs: environment API proposal (#16089)
vitejs/vite: docs/guide/api-vite-environment.md

`onUpdate` is called only once when the new runtime is initiated. It passed down a method that should be called when connection triggers the HMR event. The implementation depends on the type of connection (as an example, it can be `WebSocket`/`EventEmitter`/`MessageChannel`), but it usually looks something like this:

```js
function onUpdate(callback) {
  this.connection.on('hmr', (event) => callback(event.data))
}
```

The callback is queued and it will wait for the current update to be resolved before processing the next update. Unlike the browser implementation, HMR updates in Vite Runtime wait until all listeners (like, `vite:beforeUpdate`/`vite:beforeFullReload`) are finished before updating the modules.

## Backward Compatibility

The current Vite server API will be deprecated but keep working during the next major.

- `server.transformRequest(url)` <br> -> `server.environment('browser').transformRequest(url)`
- `server.transformRequest(url, { ssr: true })` <br> -> `server.environment('ssr').tranformRequest(url)`
- `server.warmupRequest(url)` <br> -> `server.environment('browser').warmupRequest(url)`

> - `server.ssrLoadModule(url)` <br> -> `server.ssrModuleRunner.executeEntryPoint(url)`

Reviewer: Shouldn't it be `server.environment("ssr").run(url)`?

Reviewer 2: I thought that `run(url)` would not return a Module instance, as in other environments it isn't possible to do that, no? `run(url)` has `void` return to me.

Reviewer: Ah, you are right

In its simpler form, the vite builder could call each build in series as defined by `config.environments` order (or by a new `config.build.environments` order). This should cover most current use cases out of the box, given that most frameworks build the client first and then SSR (as they use the client manifest).

It is an interesting design space to explore because it could make build and dev work in a more uniform way. For example, Vite Builder could also only load the config file once and do the overrides for each environment in the exact same way as it is done during dev. And there could also be a single plugin pipeline, if we would like plugins to share state directly between the environments (as it happens during dev already).

## Backward Compatibility

The current Vite server API will be deprecated but keep working during the next major.

> - `server.transformRequest(url)` <br> -> `server.environment('browser').transformRequest(url)`

Reviewer 2: I was afraid it will end up being too wide, but it looks good. Pushed. Thanks!

<img width="756" alt="image" src="https://github.com/vitejs/vite/assets/583075/12f01786-bd07-4966-b22a-f75318cf2491">

```

Each environment can also be accessed through its name:

```js
server.environment('browser').transformRequest(url)
```

An environment is an instance of the `ModuleExecutionEnvironment` class:

```ts
class ModuleExecutionEnvironment {
  /**
   * Unique identifier for the environment in a Vite server.
   * By default Vite exposes 'browser' and 'ssr' environments.
   * The ecosystem has concensus on other environments, like 'workerd'.
   */
  name: string
  /**
   * Communication channel to send and receive messages from the
   * associated module runner in the target runtime.
   */
  hot: HMRChannel | null
  /**
   * Graph of module nodes, with the imported relationship between
   * processed modules and the cached result of the processed code.
   */
  moduleGraph: ModuleGraph
  /**
   * Trigger the execution of a module using the associated module runner
   * in the target runtime.
   */
  run: ModuleRunFunction
  /**
   * Resolved config options for this environment. Options at the server
   * global scope are taken as defaults for all enviroments, and can
   * be overriden (resolve conditions, external, optimizedDeps)
   */
  config: ResolvedEnvironmentConfig

>   constructor({ name, hot, run, config }: ModuleEnvironmentOptions)

Reviewer: Should it receive a server? How does it access `pluginContainer` in `transformRequest`?

Reviewer 2: Corrected 👍🏼

        true
      )
    }
    return []
  }
  ```

- Return an empty array and perform complete custom HMR handling by sending custom events to the client:

  ```js
  hotUpdate({ environment }) {
    if (environment !== 'browser')
      return

    server.environment(environment).hot.send({
      type: 'custom',
      event: 'special-update',
      data: {}
    })
    return []
  }
  ```

  Client code should register corresponding handler using the [HMR API](./api-hmr) (this could be injected by the same plugin's `transform` hook):

  ```js
  if (import.meta.hot) {
    import.meta.hot.on('special-update', (data) => {
      // perform custom update
    })
  }
  ```

### Using `apply`

In the examples of the previous section, we used a guard in the `hotUpdate` hook to only process updates from the browser environment. If a plugin is specific to only some of the available environments, `apply` can be used to avoid guarding each hook and improve performance.

```js
function ssrOnlyPlugin() {
  return {
    name: 'ssr-only-plugin',
    apply({ environment }) => environment === 'ssr',
    // unguarded hooks...
  }
}
```

### Accessing the current environment in hooks

> The `environment` instance is passed as a parameter to the `options` of `resolveId`, `load`, and `transform`. It can also be accessed using the plugin hook context by `this.environment`.

Reviewer: Is `environment` instance a string?

# Vite Environment API

:::warning Low-level API

> Initial work for this API was introduced in Vite 5.1 with the name [Vite Runtime API](./api-vite-runtime.md). In Vite 5.2 the API was reviewed and renamed as Vite Environemnt API. It remains an experimental feature. We are gathering feedback about the revised proposal [here](https://github.com/vitejs/vite/discussions/15774). There will probably be breaking changes to it in Vite 5.3, so make sure to pin the Vite version to `~5.2.0` when using it. This is a low-level API meant for library and framework authors. If your goal is to create an application, make sure to check out the higher-level SSR plugins and tools at [Awesome Vite SSR section](https://github.com/vitejs/awesome-vite#ssr) first.

Reviewer 3: ```suggestion
Initial work for this API was introduced in Vite 5.1 with the name [Vite Runtime API](./api-vite-runtime.md). In Vite 5.2 the API was reviewed and renamed as Vite Environment API. It remains an experimental feature. We are gathering feedback about the revised proposal [here](https://github.com/vitejs/vite/discussions/15774). There will probably be breaking changes to it in Vite 5.3, so make sure to pin the Vite version to `~5.2.0` when using it. This is a low-level API meant for library and framework authors. If your goal is to create an application, make sure to check out the higher-level SSR plugins and tools at [Awesome Vite SSR section](https://github.com/vitejs/awesome-vite#ssr) first.
```

All these environments share Vite's HTTP server, middlewares, and Web Socket. The resolved config and plugins pipeline are also shared, but plugins can use `apply` so its hooks are only called for certain environments. The environment can also be accessed inside hooks for fine-grained control.

![Vite Environments](../images/vite-environments.svg)

## Using environments in the Vite server

A Vite dev server exposes two environments by default named `'browser'` and `'ssr'`. The browser environment runs in client apps that have imported the `/@vite/client` module. The SSR environment runs in the same runtime as the Vite server (for example, in node) and allows application servers to be used to render requests during dev with full HMR support. We'll discuss later how frameworks and users can create and register new environments.

The available environments can be accessed using the `server.environments` read-only array:

```js
server.environments.forEach((environment) => log(environment.name))
```

Each environment can also be accessed through its name:

```js
server.environment('browser').transformRequest(url)

> ```

Reviewer 4: I think the shape of `server.environments` is confusing. It looks like a function and an array at the same time. I think we need to decide if it should be an array, Set, or an object shape. Maybe `server.environment('browser')` can be `server.environments.get('browser')`? It seems like the goal is to prevent mutations.

Reviewer 2: A map would be great, but we wanted an array because we need to have an order to call `hotUpdate` for example, and also to build the environments if we decide to get into that game.

But we could have that order in a different way. We would need these to be in sync but maybe just an array of names is ok here, also for configuration (`server.hmr.environments: ['ssr', 'browser']`, `build.environments: ['browser', 'ssr']`). I think some folks wanted SSR first for HMR, but most frameworks build the client first.

It would be fine for me to switch to a map and `server.environments.get('browser')`. But even then, I see value in offering `server.environment('browser')` as a shortcut to whatever we choose as it reads better to me.

Reviewer 2: Just an idea, not convinced this is good, but it may be interesting. We could have a function define the order instead of using arrays so users can decide to execute in parallel too if their frameworks are fine with that:
```js
// Access
server.environment('name')

// Adding new environments using a plugin hook
registerEnvironments(server, environments: Set<Environment>) {
  // ...
  environments.add(workerdEnvironment)
}

// Define order for `hotUpdate` using a config option
server.modulesChange = (server, handleHotUpdate: (environment: string) => Promise<void> {
  // Orchestrate handling of hot updates in series
  await handleHotUpdate('ssr')
  await handleHotUpdate('browser')

  // or do it in parallel
}

// Define order for build time using a config option
build.defineBuilds = (builder, build: (environment: string) => Promise<void>) {
  // Orchestrate builds
  await build('browser')
  await build('ssr')

  // or do it in parallel
}
```
I think we would still need a way to query all environments, as in `server.environments`. Maybe we should have environments be a Map<string, environment> as proposed before.

---

Something else about a Map/Set vs Array that we discussed with Vladimir. Using an array and trying to avoid the use of names as a unique identifier, would potentially allow having two environments of the same type. Vlad was thinking about Vitest monorepo support and how he could use a JSDom environment for each package. It isn't clear to me or him if this is really needed. Just sharing as we could try only using instances to avoid referring to them by name. Or maybe we need an environment `id` and a `name`. So the id is unique `{package-name}:jsdom` and the name is not `jsdom`.

Reviewer 4: I'm not sure how ordered environments will work yet, as you mentioned different frameworks have different needs. The ones I'm aware of are:

- SvelteKit: Runs `server`, `client`, then `worker` (service worker, optional) builds.
- Astro: Runs `server` then `client` builds.

But even so, I don't think we should make the order of environments matter. All environments tasks like `hotUpdate` should kick off at once. It would be weird if the `ssr` HMR have to wait for the `browser` HMR to finish for example. We also wouldn't want Vite plugins to have make assumptions of the order too.

I think we should tackle the environment builds concept later, but the idea of annotating the "type" of build this way is interesting.

---

Re `server.environments`, it's not clear to me what we want to expose according to the docs. I'm not sure I can suggest something correct without missing a goal you have 😅 I was thinking we're exposing a plain object that only has the `forEach` and `get` keys.

Reviewer 2: I'm liking more and more the idea of using a config function to orchestrate builds and HMR propagation. It is easy to maintain and implement for us, and would let frameworks do whatever they want (series vs parallel, or a mix of them, and add logic between builds if needed too without relying on `buildEnd`).

About HMR propagation order for environments, we got a request for this feature here: https://github.com/vitejs/vite/discussions/15774#discussioncomment-8575835. I also coded it as parallel as you suggest in the separate module graphs PR. We can have it in parallel by default and only add a config option function to orchestrate if HMR should be in series or parallel, or a mix, when we have good use cases (if the request I linked is not enough for that already).

---

I think the requirements for environments are:
- Quickly get an environment given its name (or id if we go there)
- Get a list of all environments (or expose forEach) so plugins can check if an environment is there already and add it if not. And to do things like logging or any operation on all environments when you don't know exactly what environments are configured.
- Register an environment. Done in a hook in the proposal, outside of the data structure.

Reviewer: > It would be weird if the ssr HMR have to wait for the browser HMR to finish for example.

Can you explain why this is weird? Folks from Vike explained that then need to resolve ssr hmr first. Right now they are already doing this by calling it inside handleHotUpdate before returning new modules. If we can provide a built-in solution, that would be great

Reviewer 4: It's weird because two environments may have delayed HMR feedback if one relies on the other. It feels brittle if HMR has to rely on a specific order of an environment.

Reviewer 2: @nitedani would you explain why you need to wait for HMR propagation to be done in SSR before starting with the browser propagation?

Reviewer 5: [I'm restarting the server runtime child process when the server code changes.](https://github.com/vikejs/vike/pull/1513/files#diff-e793b1e19f4b3aaf303c35365faf66c66bf3bd9da518311fb2a991eea715f3a7R51) If the browser reload doesn't wait for the server restart, the server might exit/restart during a request/response, causing an error page in the browser. If the browser is reloaded only after the server restart, it stays in loading state until the [server is listening](https://github.com/vikejs/vike/pull/1513/files#diff-6b65b38849e385354eaa6b29d3bca0954a5923549eba16a9415be870b1e1fbcdR28).

Reviewer 6: Remix would _love_ if Vite had first class support for ordering/orchestrating builds. We wrote a bit about why [Remix couldn't just be a Vite plugin](https://remix.run/blog/remix-vite-stable#more-than-just-a-vite-plugin) anymore.

Here's some relevant code where we run the [client build first and then run N server builds](https://github.com/remix-run/remix/blob/main/packages/remix-dev/vite/build.ts#L288-L316) for our server bundles feature (N defaults to 1).

Reviewer 2: For reference, there are many frameworks in the same situation and there has been several attempts to move orchestration of builds into core in the past:
- https://github.com/vitejs/vite/issues/5936
- https://github.com/vitejs/vite/discussions/9496

We discussed this quite a lot in the team, and the last consensus was that Vite didn't want to get into that game. See [comment from Anthony](https://github.com/vitejs/vite/discussions/9496#discussioncomment-3384709) summarising the position of the team at that point.

This didn't prevented most of the frameworks to hack their way into being a plugin because the benefits outweighed the hacks for them (like starting a new vite build at in the `writeBundle` hook of the prev build). SvelteKit, SolidStart, Remix, Qwik City, Vike, and a lot of others are in this same situation.

Looking back, I think it was maybe a good idea to avoid making this request a core feature at that point. We didn't have the concepts in place to go there first. I think that if we have a proper Environment high-level concept now, the position of several team members will be different and there is a better chance for `vite build` to take more responsabilities.

Reviewer 7: Will Rolldown transpile the server- and client-side at the same time?

This question is one reason why I'm inclined to think that it's best if Vite owns the sequencing of the build. In other words, that it should be Vite's responsibility to determine what happens when `$ vite build` is run while introducing new options such as `build.parallel: boolean`.

Another small rationale is [`ssrEmitAssets`](https://github.com/vitejs/vite/pull/11430) which every SSR framework needs to use and implement. If Vite owns the build sequencing then Vite could implement that on behalf of SSR frameworks.

<hr>

That said, in order to enable plugins such `vite-plugin-pwa`, `@vitejs/plugin-legacy` and the potentially upcoming hosting Vite plugins (Cloudlare/Netlify/Vercel/...), it may make sense if Vite provides an official way to hook into the build process. I guess whether this is needed depends on how the Vite Environment API evolves.

<hr>

@pcattori I can see the marketing appeal of it, but I recommend Remix against doing that. In my experience (Vike aka vite-plugin-ssr was amongst the first, if not the first, framework to do that), (mis-)using the Vite CLI is causing issues for zero technical gain. The only gain I can see is a marketing one. [Quoting antfu](https://github.com/vitejs/vite/discussions/9496#discussioncomment-3384709):

> We recommend higher-level integrations to provide their own CLI to fit various needs. For example, a framework could proxy other Vite subcommands but only provide custom logic for its build command.

I agree.

Simple examples:
 - Imagine you want to have a new CLI flag to disable pre-rendering: it's confusing for the user that `$ vite build --disable-pre-rendering` is a framework-specific CLI argument.
 - Imagine you want to introduce a new CLI command that lists all the app's routes. It's confusing for the user to use `$ vite routes`  that is framework specific, or having to use two CLIs `$ vite` and `$ remix`.

At the end of the day I think it boils down that: beyond the marketing aspect of it, is there any benefit of (mis-)using Vite's CLI?

Funny coincidence: using a dedicated Vike CLI instead of (mis-)using Vite's CLI is actually the next big task in Vike's pipeline.

Reviewer 7: - Cannot (or hacky / weird for the user) extend new CLI commands & arguments. (See the two examples of my previous reply.)
- [Need to duplicate and maintain a copy of all Vite CLI options](https://github.com/vikejs/vike/blob/b65c343ddb6da0904f16970e0d8b0091830314b4/vike/node/plugin/plugins/autoFullBuild.ts#L75).
- Vite version mismatch. (Vite CLI resolves to the user-land `node_modules/vite/`, whereas the Vite API resolves to Vike-land `node_modules/.pnpm/vike/node_modules/vite/`).
- [Vike specific] Vike now directly uses esbuild to transpile Vike config files. Vike needs to resolves its config before spinning up the Vite instance. It's a hard requirement for the upcoming Vike versions, thus Vike cannot use the Vite CLI anymore.
- [Vike specific][Experimental] Vike is going to forgo the entire concept of CLI arguments and instead go for `$ VIKE={prerender:true,ssr:false} vike build`.

An environment is an instance of the `ModuleExecutionEnvironment` class:

```ts
class ModuleExecutionEnvironment {
  /**
   * Unique identifier for the environment in a Vite server.
   * By default Vite exposes 'browser' and 'ssr' environments.
   * The ecosystem has consensus on other environments, like 'workerd'.
   */
  name: string
  /**
   * Communication channel to send and receive messages from the
   * associated module runner in the target runtime.
   */
  hot: HMRChannel | null
  /**
   * Graph of module nodes, with the imported relationship between
   * processed modules and the cached result of the processed code.
   */
  moduleGraph: ModuleGraph
  /**
   * Trigger the execution of a module using the associated module runner
   * in the target runtime.
   */

>   run: ModuleRunFunction

Reviewer 8: What is the type of `ModuleRunFunction`? Is it `(...args: any) => any` or `() => void`?

Reviewer 2: So far, the proposal assumed it to be `(url: string) => void`. But given your questions, and a discussion between @sheremet-va and @bluwy yesterday, I think it may be better for it to be `(url: string, options?: any) => any` and let each environment define the options and return type so code like you proposed [here](https://github.com/vitejs/vite/pull/16089#discussion_r1512155384) will work:
```js
// the input and the output depends on the environment
const response = await server.environment('workerd').run(nodeReqToNativeReq(req))
const html = getResponseContent(response)
```

Reviewer 2: I forgot we wanted to allow for passing the request too. I think it could go inside options for environments that would need that, no?

Reviewer: Options are too environment-specific, we need a general interface. I would expect most environments to throw an error if request is jot passed

Reviewer 8: > we need a general interface that frameworks can just use without knowing what environment it is.

To achieve this, I guess we need to specify the return type. For example, limiting to a `Response` object.

Reviewer: I don't think we can limit the return type to be honest. I would make it unknown so you have to check the return, but you can always _call_ it with same parameters at least.

Reviewer 9: I'm a bit confused about this and It looks like this is the only interface for "environment implementer" (e.g. cloudflare workerd) to provide a unique feature to "environment users" (i.e. framework authors).

I feel this interface is somewhat analogical to what @sapphi-red's suggested to use `plugin.api` in
https://github.com/sapphi-red/vite-envs/blob/7f76892b7d28f0da06826f43953cedb5b2f042c5/packages/example-framework/src/index.ts#L16-L18C19

So, how about Vite to only define this interface with no constraint like this?

```ts
interface ModuleExecutionEnvironment {
  ...
  api?: unkonwn
}
```

Then, environment implementer side might provide:

```ts
interface WorkerdEnvironmentApi {
  dispatchFetch(req: Request): Promise<Response>
}
```

and framework authors would use it in a similar way as sapphi-red's `plugin.api` (or also Cloudflare team's POC, which is likely inspired by sapphi-red's api https://github.com/dario-piotrowicz/vite-runtime-5.1-experimentations/blob/2fa5ecead1fc60b30a4952f8bd7fdf4bf22aba4f/examples/example-framework/vite.config.js#L9-L15)

```ts
import type { WorkerdEnvironmentApi } from "vite-environment-workerd"

const frameowkrPlugin = {
  configureServer(server) {
    const workerdEnv = viteServer.getEnvironment("workerd");
    assert(workerdEnv);
    const workerdApi = workerdEnv.api as WorkerdEnvironmentApi;
    return async () => {
      server.middlewares.use(async (req, res) => {
        const webRes = await workerdApi.dispatchFetch(toWebRequest(req));
        fromWebRequest(webRes, res)
      }
    }
  }
}
```

Reviewer 2: For reference, we are also discussing with @IgorMinar in this thread about `run`: https://github.com/vitejs/vite/pull/16089#discussion_r1515566271. Commented about the serialization part there.

Would you expand on the environment initialization issue?

I think how `run` (or something like your `api`) proposal will pan out is the biggest unknown to be defined right now in the proposal.

  /**
   * Resolved config options for this environment. Options at the server
   * global scope are taken as defaults for all environments, and can
   * be overridden (resolve conditions, external, optimizedDeps)
   */
  config: ResolvedEnvironmentConfig

  constructor(
    server,

>     { name, hot, run, config }: ModuleExecutionEnvironmentOptions,

Reviewer 2: `run` here isn't the module runner but a function that connects the ModuleExecutionEnvironment instance to the ModuleRunner (possibly in another runtime as you said). This function will use function calls or RPC to be able to execute code in the runner and get the result.

The `run` function at first was a way to trigger execution, but given that we want to abstract the `ssrEnvironment`, we need to also be able to get the results and interact with them (your [comment here](https://github.com/vitejs/vite/pull/16089#discussion_r1515462163)). @sheremet-va had some ideas that we still need to explore and write down, that if I understand correctly would allow us to avoid the need for environments to specify `run` but rather implement some more low level transport layer so we can implement it in Vite core directly.

  )

  /**
   * Resolve the URL to an id, load it, and process the code using the
   * plugins pipeline. The module graph is also updated.
   */

>   async transformRequest(url: string)

Reviewer 2: I'll send a commit soon. The result is:
```ts
export interface TransformResult {
  code: string
  map: SourceMap | { mappings: '' } | null
  etag?: string
  deps?: string[]
  dynamicDeps?: string[]
}
```

  /**
   * Register a request to be processed with low priority. This is useful
   * to avoid waterfalls. The Vite server has information about the imported
   * modules by other requests, so it can warmup the module graph so the
   * modules are already processed when they are requested.
   */
  async warmupRequest(url: string)

  /**
   * Fetch information about a module from the module runner without running it.
   */

>   async fetch(url: string)

Reviewer 2: @sheremet-va maybe it could be something like `fetchModuleInfo(url)` or `getModuleInfo(url)`?

>   async fetch(url: string)

Reviewer: Ideally, if we expose `run` function and don't require it in options, then this method should not even be exposed

```js
app.use('*', async (req, res, next) => {
  const url = req.originalUrl

  // 1. Read index.html
  let template = fs.readFileSync(path.resolve(__dirname, 'index.html'), 'utf-8')

  // 2. Apply Vite HTML transforms. This injects the Vite HMR client,
  //    and also applies HTML transforms from Vite plugins, e.g. global
  //    preambles from @vitejs/plugin-react
  template = await server.transformIndexHtml(url, template)

  // 3. Load the server entry. executeEntryPoint(url) automatically transforms
  //    ESM source code to be usable in Node.js! There is no bundling
  //    required, and provides full HMR support.
  const { render } = await server.ssrModuleRunner.executeEntryPoint(
    '/src/entry-server.js',

>   )

Reviewer 8: If a user wants to run a code in workerd, would the code be like below?
```js
// the input and the output depends on the environment
const response = await server.environment('workerd').run(nodeReqToNativeReq(req))
const html = getResponseContent(response)
```

Reviewer 2: Let's merge this thread into https://github.com/vitejs/vite/pull/16089#discussion_r1512147458

  // 4. render the app HTML. This assumes entry-server.js's exported
  //     `render` function calls appropriate framework SSR APIs,
  //    e.g. ReactDOMServer.renderToString()
  const appHtml = await render(url)

  // 5. Inject the app-rendered HTML into the template.
  const html = template.replace(`<!--ssr-outlet-->`, appHtml)

  // 6. Send the rendered HTML back.
  res.status(200).set({ 'Content-Type': 'text/html' }).end(html)
})
```

## Environment agnostic SSR

One of the objectives of this proposal is to allow users to swap Vite's default SSR environment. Out-of-the-box, Vite will use a node environment running on the same process as the server. But a user may want to swap it for a workerd environment.

Vite would expose a `server.ssrEnvironment` so all frameworks can use it and allow them to define what environment type should be used for it. Ideally, instead of using `server.nodeModuleRunner`, the example above would be written as:

```js
// 3. Load the server entry, with full HMR support.
const { render } = await server.ssrEnvironment.run('/src/entry-server.js')

// 4. render the app HTML.
const appHtml = await render(url)
```

> `run(url)` would return RPC wrappers for `entry-server` exported functions, so that this code is compatible with both node and workerd environments.

Reviewer 2: This is very interesting. I'm leaning towards having good error messages here, rather than restricting what the user can do with the Node Environment. At least for Astro and SvelteKit I know they will be more than happy to review and fix their code as soon as they see they can get workerd SSR in dev by doing so. And we may be able to work with the authors of popular plugins too.

But you have a good point. Maybe there is a way to warn users and force them to disable the warning for example? My only worry is that there may be cases where Vite is used to develop tooling and not a framework that is only intended to work in Node and I don't think we should restrict them in that case.

We could have a "safe" mode where we go through serialization even when using the node environment to make sure it will work everywhere.

> `run(url)` would return RPC wrappers for `entry-server` exported functions, so that this code is compatible with both node and workerd environments.

Reviewer: I agree that `run` function should always enforce serializable data. For loading Node-SSR code, frameworks can always use `nodeRunner` or something maybe 🤷🏻 I am also ok with exposing an option to enable enforcement.

Another point to consider is if we even want to expose the module. Currently, a lot of frameworks use `ssrLoadModule` and call methods freely, but with the new API, they would _have_ to refactor their code, so maybe it's better to limit this function to simply running the module itself, and not return any RPC layer. On the other hand, I see how it can be useful to expose some methods knowing the limitations to establish routes for example:

```js
const { startRouter } = await environment.run('/entry.js')
await startRouter(['/api', '/api/puppies'])
```

The alternative would be to run separate modules:

```js
await environment.run('/entry.js?routes=/api,/api/puppies')
// calls "startRouter" inside the module
```

> `run(url)` would return RPC wrappers for `entry-server` exported functions, so that this code is compatible with both node and workerd environments.

Reviewer 9: Coming from the comment https://github.com/vitejs/vite/pull/16089#discussion_r1515592481 and especially for Workerd integration, I think loading "entrypoint file" is a part of Miniflare environment initialization, so the API such as `run: (entyFile: string) => Promise<anyModule>` didn't seem easy to provide based on the current approach used by experiments so far. 

This is not what I'm familiar with, but maybe durable objects would be a good fit to implement such "proxy module"?

> `run(url)` would return RPC wrappers for `entry-server` exported functions, so that this code is compatible with both node and workerd environments.

Reviewer 2: Added a clarification after internal discussions (it still need a better explanation):  https://github.com/vitejs/vite/pull/16089/commits/650106656b0aa8dcb45fc8a277a5b5c796f5dcf4

We are thinking about separating the RPC layer into a feature we could deliver independently. If there is demand and we find a good way to do it, I think it would be very helpful to make environments easier to use but it isn't strictly needed.

So it is being proposed that the first MVP will only have `environment.run(url): void`.

And we could later on (if we find a good way to do it), a `environment.import(url): RPC exports` (or `load`?). According to @IgorMinar they are seeing that a lot of frameworks end up re-implementing a RPC scheme, so there could be value in sharing it at the Vite core level. And I like that it makes the node environment and the workerd environment a lot easier to swap, given that we help users avoid non-serializable data by having opt-in or opt-out guards against that. @sheremet-va mention yesterday that if users want to use non-serializable data, they could directly use the node runner:
```js
// RPC export, even for the node environment, can't be opt-out, we ensure that
// the environment can be swapped by any other environment that is able to 
// implement the RPC `import` function.
const { render } = environment.import(url)

// Equivalent to `ssrLoadModule`, there is no restriction here
const { render } = moduleRunner.import(url)

// The node module runner is exposed as `server.nodeModuleRunner` and is part of 
// Vite's default toolbox when you create a dev server.
const { render } = server.nodeModuleRunner.import(url)

// But you won't be using it for SSR, as we want to allow swappable environments,
// so you'll use the exposed `server.ssrEnvironment` that frameworks define that works
// on all environments that implemented RPC exports for import
const { render } = server.ssrEnvironment.import(url)

// We should only recommend this for SSR if we find a good way to make RPC work
// in workerd. Until then, we would propose using `environment.run(url): void`
server.ssrEnvironment.run(url)
// and so on... 
```

> `run(url)` would return RPC wrappers for `entry-server` exported functions, so that this code is compatible with both node and workerd environments.

Reviewer 8: If we are first going only with `run`, I guess we need to expose something like `transformIndexHtml` on ModuleRunner.
```ts
// in node world

server.ssrEnvironment.run('/entry.js')
```
```ts
// entry.js running in other runtime (e.g. workerd)

const templatePath = '/index.html'

export default async (req: Request) => {
  const template = await moduleRunner.transformIndexHtml(req.url, templatePath) // this function

  const renderer = await moduleRunner.executeEntrypoint()
  const appHtml = renderer(req.url)

  const html = template.replace(`<!--ssr-outlet-->`, appHtml)

  return new Response(html, { headers: { "content-type": "text/html;charset=UTF-8" } })
}
```
Also I guess going this way is not compatible with plugins using `configureServer` hook to inject middlewares unless they make a middleware that proxies requests from Vite server to the other server.

> `run(url)` would return RPC wrappers for `entry-server` exported functions, so that this code is compatible with both node and workerd environments.

Reviewer: We can brainstorm some kind of a solution. Maybe returning a string is fine? Or do we require a Response? Should we push standard Web APIs here? We can use this opportunity to refine the SSR API that relies on running code on another server, and keep `ssrLoadModule` as a separate thing just for running code on the server

> `run(url)` would return RPC wrappers for `entry-server` exported functions, so that this code is compatible with both node and workerd environments.

Reviewer: > I guess we need to expose something like `transformIndexHtml` on ModuleRunner.

Also, can't you just use it as a virtual module? 🤔 It should just return a string, right? So in the workerd it's something like this:

```ts
const html = import.meta.env.DEV ? await import(`virtual:ssr-html/${req.url}`).default : await import('/index.html').default
return new Response(html.replace('<!-- ssr -->', reactRenderer.renderToString(<App>)))
```

> `run(url)` would return RPC wrappers for `entry-server` exported functions, so that this code is compatible with both node and workerd environments.

Reviewer 8: Ah, that's true.

## Plugins and environments

The Vite server has a shared plugin pipeline, but when a module is processed it is always done in the context of a given environment.

### The `hotUpdate` hook

- **Type:** `(ctx: HotContext) => Array<ModuleNode> | void | Promise<Array<ModuleNode> | void>`

> - **See also:** [HMR API](./api-hmr)

Reviewer 4: I don't quite understand why we need to introduce a new hook for this. Couldn't we support something like:

```js
{
  name: 'my-plugin',
  handleHotUpdate: {
    environment: true,
    handle(...) { ... }
  }
}
```

`hotUpdate` doesn't quite follow the convention like `resolveId`, `load`, `renderChunk` I think (verb+noun), especially for hooks that return something. `hotUpdate` sounds like `buildStart`, `renderStart`, which are hooks for certain events that don't return anything.

Reviewer 2: We could add `environment: true` but I think it is better to have a clean hook for the future. At some point, everyone will migrate to environments if we move forward and it is good we don't need to use the object form forever. `handleHotUpdate` as a function will be removed as it is working with mixed modules.

I agree that `hotUpdate` may not be the best name. Maybe we could use something similar to Rollup's `watchChange`? I think `modulesChange` may be a good name. We can brainstorm other ideas, `modifyHotUpdate` may be a good description of what the hook does too

Reviewer 4: Isn't `watchChange` like `renderStart` where they are "events" and the hooks don't return anything? I'm not sure why we have to rename `handleHotUpdate`, I think it's fine to leave it as is, especially that the type signature looks very similar to before.

Reviewer 2: We need to rename it because `handleHotUpdate` receives a context with mixed module nodes (browser and ssr), and we need to keep that for backward compatibility. It is also called only once with module nodes from both graphs. The new function looks similar but it is called once per environment on file change, and gets a context with modules that only correspond to that environment.
As far as I can see, if we don't have a new hook here, we can't properly get backward compatibility.

Does `modifyHotUpdate` sounds better to you?

Reviewer 4: Yeah I also noticed those differences too and it doesn't feel big to me. We can use the object form if they want to opt to a more fine-grained env-specific handling. `modifyHotUpdate` sounds odd to me 😅

Reviewer 2: But it isn't about opting in a more fine-grained env-specific handling. The idea is that is going to be the only way to handle hot updates. We don't want to keep the mixed module graph and all the compat code forever. Ideally that all that is deleted in Vite 7. Because we can't change `handleHotUpdate` for back compat, we'll need to use the object form with a flag if we don't want a new name. And that means that in the future this will be the only special hook that you can only use with a flag and object form and can't use as a normal hook. I think it is a lot easier to just have a new hook. We can keep brainstorming a name for it until we have something that clicks for you. Maybe `controlHotUpdate`, `handleHotEvent`, `modifyHotModules`, `handleHotPropagation`, `modifyHotPropagation`.

Reviewer 4: Yeah I see your point too. I guess on paper it's kinda unfortunate we have to rename it while looking very similar. But if we do want a different name so it's easier to migrate, `handleHotModules` and `handleHotPropagation` sounds nice. It's often mistaken that `handleHotUpdate` generally handles hot updates, but actually it's a hook that filters the modules to be propagated for hmr.

    for (const mod of modules) {
      serverEnvironment.moduleGraph.invalidateModule(
        mod,
        invalidatedModules,
        timestamp,
        true
      )
    }
    return []
  }
  ```

- Return an empty array and perform complete custom HMR handling by sending custom events to the client:

  ```js
  hotUpdate({ environment }) {
    if (environment !== 'browser')
      return

    server.environment(environment).hot.send({
      type: 'custom',
      event: 'special-update',
      data: {}
    })
    return []
  }
  ```

  Client code should register the corresponding handler using the [HMR API](./api-hmr) (this could be injected by the same plugin's `transform` hook):

  ```js
  if (import.meta.hot) {
    import.meta.hot.on('special-update', (data) => {
      // perform custom update
    })
  }
  ```

### Using `apply`

In the examples of the previous section, we used a guard in the `hotUpdate` hook to only process updates from the browser environment. If a plugin is specific to only some of the available environments, `apply` can be used to avoid guarding each hook and improve performance.

```js
function nodeOnlyPlugin() {
  return {
    name: 'node-only-plugin',

>     apply({ environment }) => environment === 'node',

Reviewer 3: Did you mean: 
```suggestion
    apply: ({ environment }) => environment === 'node',
```

    // unguarded hooks...
  }
}
```

### Accessing the current environment in hooks

> The `environment` name is passed as a parameter to the `options` of `resolveId`, `load`, and `transform`.

Reviewer 8: If we want to conditionally return a virtual file for worker-like runtimes such as  [Cloudflare workerd](https://github.com/cloudflare/workerd) and [Fastly Compute](https://www.fastly.com/documentation/guides/compute/javascript/), is there a way to do it at once?

I guess it is required to write it like:
```js
const p = {
  name: 'virtual',
  load(id, opts) {
    // it requires to write all the environment names
    if (opts.environment === 'workerd' || opts.environment === 'fastly-compute') {
      return 'a code for worker-like runtimes'
    }
    return 'a code'
  }
}
```
But to migrate from `config.ssr.target === 'webworker'`, I think it'd nice to write it something like:
```js
const p = {
  name: 'virtual',
  load(id, opts) {
    const e = server.environment(environment)
    if (e.type === 'worker') {
      return 'a code for worker-like runtimes'
    }
    return 'a code'
  }
}
```

Reviewer 2: This is a great point. We discussed two weeks ago a bit about this with @sheremet-va. In Vite core plugins, we have several conditions that depend on the `ssr` flag and other that depend on `ssr.target: node | webworker` as you explain (other `ssr` config options are already accounted for in the current state of the proposal).

If we want to remove the `ssr` flag, we need something like you propose here. I think we should move away from conditionals based on the environment name to conditionals based on what the environment supports. So, totally agree with you. I don't think it should be called `type` as it is too generic, but we need to identify what are the conditions not covered by `resolve.conditions` et al and add them as part of the environment config.

Maybe something like:
```ts
type runtimeType = 'browser' | 'node' | 'worker' | 'edge' | ?
```

The problem with this is that there will be differences between edge environments and node environments. So it may be worth checking all conditionals in core and other projects and see if we can be even more fine-grained. Maybe it is a more generic option we should add to resolve for example.

If we end up using only fine-grained options in core, I still think that having a `runtimeType` is a good idea, even if it is only to help with debugging.

Reviewer: Can you show an example when this is useful? I think having fine-grained config should be enough.

Reviewer 2: I was thinking on the `runtimeType` also as a way for community plugins to be able to ship features before we add the proper options in Vite core. If we are missing a fine-grained config, it takes time for it to end up in a release. So it may be useful to know the runtime type to do some conditionals until they can switch to fine-grained options. I'm good avoiding this until we have a good use case to justify it.

Reviewer 8: I assumed that we have many `ssr.target` conditions in our codebase, but it turns out those can be rewrote with a fine-grained option. Yeah, it seems we don't need something like this.
I was thinking of returning a virtual module conditionally, but I noticed that it can be achieved by using an `exports` field with `wintercg` condition.

  getModulesByFile(file: string): Set<ModuleNode> | undefined

  onFileChange(file: string): void

  invalidateModule(
    mod: ModuleNode,
    seen: Set<ModuleNode> = new Set(),
    timestamp: number = Date.now(),
    isHmr: boolean = false,
  ): void

  invalidateAll(): void

  async updateModuleInfo(
    mod: ModuleNode,
    importedModules: Set<string | ModuleNode>,
    importedBindings: Map<string, Set<string>> | null,
    acceptedModules: Set<string | ModuleNode>,
    acceptedExports: Set<string> | null,
    isSelfAccepting: boolean,
  ): Promise<Set<ModuleNode> | undefined>

  async ensureEntryFromUrl(
    rawUrl: string,
    setIsSelfAccepting = true,
  ): Promise<ModuleNode>

  createFileOnlyEntry(file: string): ModuleNode

  async resolveUrl(url: string): Promise<ResolvedUrl>

  updateModuleTransformResult(
    mod: ModuleNode,
    result: TransformResult | null,
  ): void

  getModuleByEtag(etag: string): ModuleNode | undefined
}
```

## Registering environments

There is a new plugin hook called `registerEnvironment` that is called after `configResolved`:

```js
function workedPlugin() {
  return {
    name: 'vite-plugin-workerd',

>     registerEnvironment(server, environments) {

Reviewer 3: ```suggestion
    registerEnvironment: (server, environments) => {
```

      const workedEnvironment = new WorkerdEnvironment(server)
      // This environment has 'workerd' as its name, a convention agreed upon by the ecosystem
      // connect workerdEnviroment logic to its associated workerd module runner

>       environments.push(workedEnvironment)

Reviewer 3: How can this plugin also notify vite that the newly registered environment should be aliased as `server.ssrEnvironment`?

Reviewer 2: There is no defined API at this point, as it wasn't part of the original proposal.
I imagine it could be something like:
```js
configureEnvironment(server, { environments, setSsrEnvironment }) {
  const workerdEnvironment = new WorkerdEnvironment(server)
  // ...
  environments.push(workerdEnvironment)
  setSsrEnvironment(workerdEnvironment)
}
```

The available environments can be accessed using `server.environments`:

```js
server.environments.client.transformRequest(url)

console.log(server.environments.ssr.moduleGraph)
```

Normally, the current `environment` instance will be available as part of the context of the code being run so the need to access them through `server.environments` should be rare. For example, inside plugin hooks.

An dev environment is an instance of the `DevEnvironment` class:

```ts
class DevEnvironment {
  /**
   * Unique identifier for the environment in a Vite server.

>    * By default Vite exposes 'browser' and 'node' environments.

Reviewer 2: Updated, we missed this one when we moved from browser/node to client/ssr. For future reference:
- Default `server.ssrEnvironment` naming. See [rationale](https://github.com/vitejs/vite/pull/16129#issuecomment-2002334260)
- Default `server.clientEnvironment` naming. See [rationale](https://github.com/vitejs/vite/pull/16194)

    mod: ModuleNode,
    seen: Set<ModuleNode> = new Set(),
    timestamp: number = Date.now(),
    isHmr: boolean = false,
  ): void

  invalidateAll(): void

  async updateModuleInfo(
    mod: ModuleNode,
    importedModules: Set<string | ModuleNode>,
    importedBindings: Map<string, Set<string>> | null,
    acceptedModules: Set<string | ModuleNode>,
    acceptedExports: Set<string> | null,
    isSelfAccepting: boolean,
  ): Promise<Set<ModuleNode> | undefined>

  async ensureEntryFromUrl(
    rawUrl: string,
    setIsSelfAccepting = true,
  ): Promise<ModuleNode>

  createFileOnlyEntry(file: string): ModuleNode

  async resolveUrl(url: string): Promise<ResolvedUrl>

  updateModuleTransformResult(
    mod: ModuleNode,
    result: TransformResult | null,
  ): void

  getModuleByEtag(etag: string): ModuleNode | undefined
}
```

## Creating new environments

One of the goals of this feature is to provide a customizable API to process and run code. Users can create new environment types using the exposed primitives.

```ts
import { DevEnvironment } from 'vite'

function createWorkerdDevEnvironment(server: ViteDevServer, name: string, config?: DevEnvironmentConfig) {

>   const workerdModuleRunner = /* ... */

Reviewer 8: I think a runner should not be declared here. What it should be here is a thing that connects the node process and the workerd process.

Or does "runner" and "moduleRunner" have a different meaning? If then, I think we should have a explanation of that.

- "runner": Always on the node process running vite. It uses a module runner in the same process (Local Runner) or a different process to run modules (Remote Runner).
- "module runner": Might be on a different process. It connects the modules and uses the ModuleEvaluator to evaluate them.

Reviewer: This can just be removed from here. The initial idea was to introduce a proxy, but we decided against it after talking with cloudflare folks because it doesn't address the main issue of calling a server on another machine (so, the ssr primitive was wrong - it's not a module but a middleware).

I will replace it with the new `RemoteEnvironmentTransport` example

```ts
  transform(code, id) {
    console.log(this.enviroment.config.resolve.conditions)
  }
```

### Registering new environments using hooks

Plugins can add new environments in the `config` hook:

```ts
  config(config: UserConfig) {
    config.environments.rsc ??= {}
  }
```

An empty object is enough to register the environment, default values from the root level environment config.

### Configuring environment using hooks

While the `config` hook is running, the complete list of environments isn't yet known and the environments can be affected by both the default values from the root level environment config or explicitely through the `config.environments` record.
Plugins should set default values using the `config` hook. To configure each environment, they can use the new `configEnvironment` hook. This hook is called for each environment with its partially resolved config including resolution of final defaults.

```ts
  configEnvironment(name: string, config: EnvironmentConfig) {
    if (name === 'rsc') {
      config.resolve.conditions = // ...

> ```

Reviewer 2: I don't think we should promote modifying the config in `configResolved`. When this hook is called, it should be safe to assume that the config is settled and plugins can read it and cache/store what they need. If we want to do this config in another hook, it should be in the `config` hook. And we could do that without a problem. The only reason why the `configEnvironment` hook exists is to help plugins deal with the complexity of environment overrides. I'm ok removing it if we think it is redundant, but at this point I feel that it will be very helpful to avoid subtle bugs in plugins.

> ```

Reviewer 9: Is this `configEnvironment` only for custom environments or do you also want to move people to configure `name = "client"` and `name = "ssr"` cases here as well? For example,

```ts
// from this
config(config, env) {
  return {
    optimizeDeps: { ... },
    ssr: { optimizeDeps: { ...} }
  }
}

// to this?
configEnvironment(name, config) {
  if (name === "client") {
    config.dev.optimizeDeps = { ... }
  }
  if (name === "ssr") {
    config.dev.optimizeDeps = { ... }
  }
}

// or this?
config(config, env) {
  config.environments.client.dev.optimizeDeps = { ... }
  config.environments.ssr.dev.optimizeDeps = { ... }
}

// or this?
config(config, env) {
  return {
    environments: {
      client: {
        dev: {
          optimizeDeps: { ... }
        }
      }
      ssr: { ... }
    }
  }
}
```

I thought it will be mostly redundant, but after writing this sample code, maybe `configEnvironment` looks explicit and nicer even for builtin ones. But maybe this assignment form of config needs to take care config mergining inside the hook itself?

> ```

Reviewer 8: Ah, I forgot that `ResolvedConfig` is a readonly. 😅

> At this point I feel that it will be very helpful to avoid subtle bugs in plugins.

What kind of bug do you have in mind?

> ```

Reviewer 2: > Is this `configEnvironment` only for custom environments or do you also want to move people to configure `name = "client"` and `name = "ssr"` cases here as well?

Yes, `configEnvironment` will be called once per environment (including default ones).

> ```

Reviewer 9: > But maybe this assignment form of config needs to take care config merging inside the hook itself?

In my comment above, I was worrying about how merging works in `configEnvironment`, but I just realized it can return partial options and merged automatically.

```ts
    configEnvironment(name, _config, _env) {
      if (name === "ssr") {
        return {
          build: {
            rollupOptions: {
              input: {
                index: entry,
              },
            },
          },
        };
      }
    },
```

## Environments during build

Plugin hooks also receive the environment instance during build. This replaces the `ssr` boolean we have been passing them so far.

In the CLI, calling `vite build` will build the Client. It is equivalent to calling `vite build --environment=client`.

The build the SSR server, the `--ssr` shourcut can be used: `vite build --ssr`. This is equivalent to calling `vite build --environment=ssr`.

Other non-default environments can be build using `vite build --environment=name`.

## Building all environments

Calling `vite build --all` will instantiate a `ViteBuilder` (build-time equivalent to a `ViteDevServer`) to build all configured environments for production. By default the build of environments is run in series respecting the order of the `environments` array. A framework or user can further configure how the build tasks are excecuted using:

```js
export default {
  builder: {
    runBuildTasks: asnyc (builder, buildTasks) => {

>       return Promise.all(buildTasks.map( task => task.run() ))

Reviewer 2: Right now, the implementation resolves the config for each build generating a separate pipeline. I did it this way so we can parallelize without issues given the current state of plugins in the ecosystem.

@antfu proposed that we should take the opportunity and try to align with dev, and have a single plugins pipeline during build so it is easier to share information during build time too. We're going to push everyone to use something like `this.environment.cache` if they need to store things while processing (maybe we need a `this.environment.api` for comms?), if the plugin isn't environment agnostic. In dev, this isn't a stretch as people are already used to do the same with the `ssr` flag (probably there are a lot of bugs due to improper isolation because things aren't as clear as having a proper environment abstraction). During build, we could do the same and ask everyone to be aware of the environment that is being built. Running the plugins in parallel, even for workers (that we may rethink as environments too) becomes possible.

I'm worried about this idea though. I think it is an interesting goal to aim for, but a lot of plugins will be broken for `vite build --all` if we do this. I was proposing that we have a `builder.isolatedPlugins` flag that we make opt-in for a major or two and let users set it if all their plugins work well in this new world.

>       return Promise.all(buildTasks.map( task => task.run() ))

Reviewer 2: About `buildStart`, we actually would like to go the other way around and call `buildStart` during dev once per environment and make plugins use that to reset the state of each environment. This could be very interesting if at one point we would like to implement a "restart this environment" if a config changes for example without the need for a full server restart.

>       return Promise.all(buildTasks.map( task => task.run() ))

Reviewer 2: It may be a bit difficult to do so given that the core parts heavily depend on the  environment abstraction that I don't know if Rollup could adopt. It is an interesting idea though, maybe once this API is stable and has proven useful, Rolldown could have environments as a first class citizen too. But this starts to blur the lines a bit of responsibilities. We always thought of Rollup/esbuild/rolldown working at a single environment level, and Vite is the abstraction that coordinates (client/ssr). So initially I'm leaning to keep that distinction.

Note: If rolldown wants to implements the dev plugin container, then it is kind of inevitable. But I'm unsure of this strategy. Maybe if we move the pluginContainer to Rust, it should still live in the Vite repo.

>       return Promise.all(buildTasks.map( task => task.run() ))

Reviewer 8: Reconsidering it, I agree with that 👍

>       return Promise.all(buildTasks.map( task => task.run() ))

Reviewer 9: Is there a way for plugin to know the start or end of certain environment build?
For example, currently it doesn't look like `environment` is accessible from `buildStart`, `buildEnd`, `writeBundle` etc...

>       return Promise.all(buildTasks.map( task => task.run() ))

Reviewer 2: We need to implement that, `this.environment` should be available for all hooks. So far we only injected it in `resolveId`, `load`, and `transform` as it was done before.
I wonder if Rollup could give us a better way to inject the environment so we don't need to wrap every hook, that feels quite inefficient

    result: TransformResult | null,
  ): void

  getModuleByEtag(etag: string): ModuleNode | undefined
}
```

## Creating new environments

One of the goals of this feature is to provide a customizable API to process and run code. Users can create new environment types using the exposed primitives.

```ts
import { DevEnvironment, RemoteEnvironmentTransport } from 'vite'

function createWorkerdDevEnvironment(server: ViteDevServer, name: string, config?: DevEnvironmentConfig) {
  const hot = /* ... */
  const connection = /* ... */
  const transport = new RemoteEnvironmentTransport({
    send: (data) => connection.send(data),
    onMessage: (listener) => connection.on('message', listener),
  })

  const workerdDevEnvironment = new DevEnvironment(server, name, {
    config: {
      resolve: { conditions: ['custom'] },
      ...config,
    },
    hot,
    runner: {
      transport,
    },
  })
  return workerdDevEnvironment
}
```

Then users can create a workerd environment to do SSR using:

```js
const ssrEnvironment = createWorkerdEnvironment(server, 'ssr')
```

## Environment Configuration

Environments are explicitely configured with the `environments` config option.

```js

> export default {
>   environments: [
>     {
>       name: 'client',
>       resolve: {
>         conditions: [], // configure the Client environment
>       },
>     },
>     {
>       name: 'ssr',
>       optimizeDeps: {}, // configure the SSR environment
>     },
>     {
>       name: 'rsc',
>       noExternal: true, // configure a custom environment
>     },
>   ],
> }

Reviewer 10: This example seems to be out of date. 🤗
```suggestion
export default {
  environments: {
    client: {
      resolve: {
        conditions: [], // configure the Client environment
      },
    },
    ssr: {
      dev: {
        optimizeDeps: {}, // configure the SSR environment
      },
    },
    rsc: {
      resolve: {
        noExternal: true, // configure a custom environment
      },
    },
  },
}
```

Reviewer 2: Thanks!

# Vite Environment API

:::warning Low-level API

> Initial work for this API was introduced in Vite 5.1 with the name "Vite Runtime API". In Vite 5.2 the API was reviewed and renamed as Vite Environment API. It remains an experimental feature. We are gathering feedback about the revised proposal [here](https://github.com/vitejs/vite/discussions/15774). There will probably be breaking changes to it in Vite 5.3, so make sure to pin the Vite version to `~5.2.0` when using it. This is a low-level API meant for library and framework authors. If your goal is to create an application, make sure to check out the higher-level SSR plugins and tools at [Awesome Vite SSR section](https://github.com/vitejs/awesome-vite#ssr) first.

Reviewer 8: ```suggestion
Initial work for this API was introduced in Vite 5.1 with the name "Vite Runtime API". In Vite 5.3 the API was reviewed and renamed as Vite Environment API. It remains an experimental feature. We are gathering feedback about the revised proposal [here](https://github.com/vitejs/vite/discussions/15774). There will probably be breaking changes to it in Vite 5.4, so make sure to pin the Vite version to `~5.3.0` when using it. This is a low-level API meant for library and framework authors. If your goal is to create an application, make sure to check out the higher-level SSR plugins and tools at [Awesome Vite SSR section](https://github.com/vitejs/awesome-vite#ssr) first.
```

Plugins should set default values using the `config` hook. To configure each environment, they can use the new `configEnvironment` hook. This hook is called for each environment with its partially resolved config including resolution of final defaults.

```ts
  configEnvironment(name: string, config: EnvironmentConfig) {
    if (name === 'rsc') {
      config.resolve.conditions = // ...
```

### The `hotUpdate` hook

- **Type:** `(ctx: HotContext) => Array<ModuleNode> | void | Promise<Array<ModuleNode> | void>`
- **See also:** [HMR API](./api-hmr)

The `hotUpdate` hook allows plugins to perform custom HMR update handling for a given environment. When a file changes, the HMR algorithm is run for each environment in series according to the order in `server.environments`, so the `hotUpdate` hook will be called multiple times. The hook receives a context object with the following signature:

```ts
interface HotContext {
  file: string
  timestamp: number
  modules: Array<ModuleNode>
  read: () => string | Promise<string>
  server: ViteDevServer
}
```

> - `this.environment` is the module execution environment where a file update is currently being processed.

Reviewer 9: Currently it's implemented as `HotContext.environment` instead of `this.environment`.
